doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -668,69 +652,83 @@ Induction
       implicitly added to this table when defined.
       Use the :cmd:`Add` and :cmd:`Remove` commands to update this set manually.
 
-.. tacn:: inversion @ident
-   :name: inversion
+:tacn:`inversion`
+~~~~~~~~~~~~~~~~~
+
+.. tacn:: inversion {| @ident | @natural } using @one_term {? in {+ @ident } }
+          inversion {| @ident | @natural } {? as @or_and_intropattern } {? in {+ @ident } }
+   :name: inversion; _
+
+   .. comment: the other inversion* tactics don't support the using clause,
+      but they should be able to, if desired.  It wouldn't make sense for
+      inversion_sigma.
+      See https://github.com/coq/coq/pull/14179#discussion_r642193096
+
+   .. todo PR: following is clear as mud
```

Comment:
> one of the hardest to explain and yet most important tactic

Then perhaps we should spend a bit more time now to get a good description.  Maybe start by identifying the use case(s) it's needed for, i.e. its purpose.  Why is deriving _all_ possible ways we could have proved the selected hypothesis useful?  And why _could have proved_?

Reply:
When performing case analysis on a hypothesis `H` whose type is an inductive with multiple constructors and non-trivial indices, there are a couple of steps:
1. Generalize any non-variable indices in `H`,
  a. optionally also generalizing those terms in the goal,
  b. optionally also generalizing those terms in all other hypotheses, and
  c. optionally generating equations that relate the fresh variables to the original terms.
2. Replace the current goal with a fresh subgoal for each constructor of the inductive type, adjusting the newly-generalized indices as necessary
  a. optionally replacing `H` in the goal with the application of the constructor to its, arguments,
  b. optionally replacing `H` in the hypotheses with the application of the constructor to its arguments,
  c. optionally leaving behind an induction hypothesis in the relevant goals
  d. optionally leaving an equation relating `H` to the application of the constructor to its arguments
3. Decompose the equations left behind by 1(c) to
  a. close absurd subgoals (such as those with an equation that `Some _ = None` or `0 = 1`) and
  b. simplify information when an equation is an equality between the same (non-dependent) constructor of an inductive type, just with different arguments (such as turning `Some x = Some y` into `x = y`)
  c. optionally doing the same simplification to dependent constructors such as those of sigma types
  d. optionally substituting any equations which equate a variable with anything else

Roughly (because I've excluded nuances such as whether things are reverted)
- `case` does 1, 1a, 2, 2a
- `case_eq` adds step 2d to `case`
- `elim` adds step 2c to `case`
- `destruct` does 1, 1a, 1b, 2, 2a, 2b; adding `eqn:` adds step 2d
- `induction` adds step 2c to `destruct`
- I think `dependent induction` adds steps 1c, 2d, and 3d to `induction`, but I haven't used it much and this is probably not quite right (because the different tactics all perform slight variations on these steps)
- `inversion` roughly performs steps 1, 1c, 2, 3a, and (roughly) a single step of 3b, unless the original inductive type was already an equality, in which case it performs 3b twice
- `inversion_clear` adds 3d to `inversion` and also clears the original `H`
- `injection` (or is it `injections`?) performs 3c repeatedly(?)
- `subst` does only 3d


Most often, `inversion` is used to perform a single step of 3c on an equality-like hypothesis.  Sometimes it's used to perform 3c on something non-equality-like, such as a vector.  Occasionally it's used for its other steps.


"all possible ways" corresponds to 2 ("each constructor"); this is the primitive building block of case analysis

"could have proved" corresponds to 3a (we don't need to consider cases that are impossible, such as the case that a list of length 1 is nil)

Reply:
@JasonGross That's a lot of detail.  It seems like it would be complex to document and for users to read.  It would take a significant effort to refine ("roughly", "probably not quite right") and get accurate (many examples/experiments).  In an ideal world, shouldn't these tactics behave more similarly with fewer one-off differences?  If so, that's an argument for less detail until we've made improvements.

Perhaps it would be helpful to identify which tactics share code (just passing different parameters).

Reply:
> In an ideal world, shouldn't these tactics behave more similarly with fewer one-off differences?

Yes.  I almost wrote that I believe that these tactics should all be replaced by a single `invert` tactic a la http://gallium.inria.fr/blog/a-new-Coq-tactic-for-inversion/

> Perhaps it would be helpful to identify which tactics share code (just passing different parameters).

Sure

Reply:
> Then perhaps we should spend a bit more time now to get a good description.

I don't feel up to the task unfortunately. That's why I summoned Jason, and from his answers, it looks to me like all we can do is acknowledge that this is all very messy and hard to understand in depth, but focus on the most important things (because from `destruct` to `inversion`, through `discriminate` and `injection`, these are all essential tactics that beginners have to learn!).

Reply:
OK, I'll make limited improvements in this PR.  For the longer term, do you think this is one of the harder points for beginners and users in general to understand/master?  And thus worth spending significant time improving the doc and/or the definition of these tactics?

Reply:
Yes, at least in my experience teaching Coq to students. Certainly worth spending time improving both the tactics and the documentation.

Reply:
[This](https://www.cs.cornell.edu/courses/cs3110/2018sp/a5/coq-tactics-cheatsheet.html#inversion) suggests that, in part, `inversion` on a hypothesis is similar to `f_equal` on a goal.

What do you think of the presentation in [Software Foundations](https://softwarefoundations.cis.upenn.edu/lf-current/IndProp.html)?  I've not dug into that yet.

In this:

![image](https://user-images.githubusercontent.com/1253341/135199865-b67474be-2df9-4f19-b36d-c5e9efecb1b9.png)

"deriving" means to add new hypotheses or new goals or both?  The number of things added is <= the number of constructors?

Reply:
Shouldn't `c i` be c with the subscript i?

Reply:
> [This](https://www.cs.cornell.edu/courses/cs3110/2018sp/a5/coq-tactics-cheatsheet.html#inversion) suggests that, in part, `inversion` on a hypothesis is similar to `f_equal` on a goal.

That's not a good description of `inversion` (it's limited to the subcases that correspond to `discriminate` / `injection`). And indeed, that's closer to how I describe `injection` to my students (doing the comparison with `f_equal`, reminding that `f_equal` can apply for any function, while `injection` only works with constructors because it uses the property that constructors are indeed *injective*).

> What do you think of the presentation in [Software Foundations](https://softwarefoundations.cis.upenn.edu/lf-current/IndProp.html)? I've not dug into that yet.

Having just a cursory look, it looks to me like a good introduction to the concept of inductive propositions and `inversion`.

> "deriving" means to add new hypotheses or new goals or both? The number of things added is <= the number of constructors?

`inversion` transforms the current goal into as many goals as there are hypotheses (except in case it can derive some contradiction for some of these goals, in which case it removes them). For each of these new goals it may add a various number of new hypotheses that gives information about the arguments to these constructors.

> Shouldn't `c i` be c with the subscript i?

Yes.

Reply:
1. Is there a simple taxonomy here?  If this is on the right track, I'll say something like this here, but may need to refine/strengthen the language:

- Inversion is recommended or necessary to prove propositions defined on inductives of type Prop.
- Inversion is generally not needed to prove propositions defined on inductives of type Set  (e.g. S (n + m) = n + (S m))
- Generally use inductively defined propositions to define subsets of other inductives (e.g. "even" is a subset of nat, as in SF)

Or maybe the key difference is the presence of the arrow in `nat → Prop` in `ev`?

```
Inductive nat : Set :=
  | O : nat
  | S : nat -> nat.

Inductive ev : nat → Prop :=
  | ev_0 : ev 0
  | ev_SS (n : nat) (H : ev n) : ev (S (S n)).
```

Maybe also be a bit more specific in the `even` example for the `Inductive` command?

2. Should this be written as `I(t)`?

![image](https://user-images.githubusercontent.com/1253341/135386305-f6b3dd66-ee10-4b8b-b57f-e54479f6abf5.png)

Reply:
Also, should the circled part be just be `t`?

![image](https://user-images.githubusercontent.com/1253341/135388980-b0f0dc7a-0cdf-4110-bf65-9211ae188ce3.png)

Reply:
> Is there a simple taxonomy here? If this is on the right track, I'll say something like this here, but may need to refine/strengthen the language:

I wouldn't present things like you do based on the sorts, but rather on whether the proposition itself has an inductive type or not. Furthermore, the proposition which should be considered for this dichotomy is not the goal but one of the hypothesis that you intend to make use of.

> Should this be written as `I(t)`?

No, I think it makes sense to use the function application notation from Coq in this context, rather than the mathematical one.

> Also, should the circled part be just be `t`?

No, the inductive type is `I` (applied to its arguments represented as `t`). If we want, we can drop the `t` part and make this just `each constructor of I`.

Reply:
Oh?  How would you define "inductive type"?  I took it to mean anything defined with `Inductive`.  Including all propositions that contains variables of that type (e.g. nat) seems so expansive here as to be meaningless.  (FWIW, I notice `True` and `False` are defined with `Inductive`, but induction and inversion are not useful for them.)

Seems to me that the constructors are defined in the `Inductive` and thus are associated with `t`, not `(I t)`.  I suppose that a proposition on `nat list` would use the constructors for `list` and not `nat`.  And who know what it would do with `(prop on nat) \/ (prop on Z)`.

Much of the difficulty in explaining inversion is that there's no explanation or example of the what the "necessary conditions" are before they're processed further, or how they are determined.

Reply:
> `inversion` transforms the current goal into as many goals as there are hypotheses (except in case it can derive some contradiction for some of these goals, in which case it removes them).

Should be "as many goals as there are constructors", right?

Reply:
See what I did.  It's probably not 100% correct.

Reply:
> How would you define "inductive type"? I took it to mean anything defined with Inductive.

You are correct.

> Including all propositions that contains variables of that type (e.g. nat) seems so expansive here as to be meaningless.

That's not what I meant. I said "whether the proposition has an inductive type" not "whether the proposition is an inductive type".

>FWIW, I notice True and False are defined with Inductive, but induction and inversion are not useful for them.

Actually, if you have `H: False` in your context, then doing `inversion H` (or `induction H` or `destruct H`) will solve the goal. If `H: True`, then `inversion H` is useless because there is no additional information to be extracted from this hypothesis (and `induction H` or `destruct H` will simply remove `H` from the context).

> Seems to me that the constructors are defined in the `Inductive` and thus are associated with `t`, not `(I t)`. I suppose that a proposition on `nat list` would use the constructors for `list` and not `nat`.

You may be confusing Coq syntax with OCaml syntax. `nat list` doesn't type check in Coq. `list nat` does. `I` is the inductive type. That's why the letter `I` was used.

>Should be "as many goals as there are constructors", right?

Yes, sorry.

Reply:
> > Seems to me that the constructors are defined in the `Inductive` and thus are associated with `t`, not `(I t)`. I suppose that a proposition on `nat list` would use the constructors for `list` and not `nat`.

> You may be confusing Coq syntax with OCaml syntax. `nat list` doesn't type check in Coq. `list nat` does. `I` is the inductive type. That's why the letter `I` was used.

nat list, list nat, that's a detail.  My point was that constructors are attached to `Inductive` definitions, i.e. associated with `t`, not with (I t).  I don't quite understand what you mean when you say "(I t) has an inductive type".

Reply:
Yes, constructors are attached to `Inductive` definition. `I` stands for `Inductive`.
I never said or suggested that `I t` has an inductive type. `I t` is *the* inductive type.

Reply:
Are you confusing `I t` with `I : t`? I fail to understand why you would think `t` is the type otherwise. In any case, I stand what I said. You can replace `I t` with just `I` if you want, but certainly not by `t`.

Reply:
We're going a bit in circles and the time delay between messages doesn't help.  Perhaps you have time to do a short video call sometime (Sunday at the earliest)?

But let's try this: suppose you have
m, n : nat
H : (m + n) = (n + m)

H is (I t).  What is t?  Isn't it `nat`?  Is it `n` or `m`?  And given an expression such H, of course it is a type, but how do you know it is an _inductive_ type?  What's the definition?

Reply:
Perhaps the issue is that in OCaml, `I t` would mean "a type family `t` with parameter `I`", while in Coq it means "an inductive type family `I` applied to the argument `t`".  In fact there's a missing nuance here, though, which is that I believe the correctly generalized case is actually `I t1 t2 ... tn`, not just `I t`.  In your example @jfehrle, `I` is the type family `eq` and `t` stands in for the sequence of arguments `nat`, `m + n`, `n + m`, I believe.  An inductive type is a type which, when reduced to head normal form, is the application of a constant defined via `Inductive` to some arguments.  This constant is called the *inductive type family*, or, if there are no arguments and the constant is already a type, it's called the *inductive type*.

Reply:
One again, `H` is not the type. The type of `H` is what's on the right side of `:`. This is what is represented by `I t` in this text, expect that, like Jason said, `I` has got more than one argument in your example.

Reply:
I think we should drop the notation `(I t)`--it's a fuzzy shorthand that AFAIK we don't use elsewhere.

Writing `I t1 t2 ... tn` isn't an improvement.  It would require explaining which of multiple Inductives would be expanded (whether n or m in my expression or whether, say,  to expand on nat or list in a different expression).  I don't know what `inversion` would do with either of those expressions.  I don't know the tactic well enough to be sure I would try the right examples--or whether they're valid.

> An inductive type is a type which, when reduced to head normal form, is the application of a constant defined via Inductive to some arguments.

What are examples of types that are _not_ inductive types?  Not sure it's useful to introduce the term at all.  Even `True` and `False` are `Inductive`s.  The thing users want to know is which `Inductive` provides the constructors.  If we know the rule and it's simple, we could state it in the doc.  And if not, we can omit that detail.

Reply:
It's `I` which provides the constructor. That's why I said that it was OK to drop the `t` and leave only the `I`.

Reply:
I think we're getting tripped up by two meanings of "type" here.  H is a name for the type `(m + n) = (n + m)` on the logic side of the Curry Howard isomorphism.  Interpreting the same symbols on the programming side, it is an expression that evaluates to a value of type `nat`.  Constructors are associated with `nat`, not with `I`.  WDYT?

Reply:
> H is a name for the type `(m + n) = (n + m)` on the logic side of the Curry Howard isomorphism. Interpreting the same symbols on the programming side, it is an expression that evaluates to a value of type `nat`.

Both sentences are incorrect. `H` is not a name for the type, it is a variable of this type. It represents a proof (i.e., a term) of this proposition / type. `H` does not evaluate to a value of type `nat`. `m` and `n` do. `H` is a value of (inductive) type `eq` (applied to arguments `(m + n)` and `(n + m)`). If you called `inversion H`, the constructors that would be used are the one for `eq`. (And BTW, since `eq` has a single constructor with no arguments, you would learn exactly nothing: the hypothesis gets just duplicated).

Reply:
Looks good to me.